### PR TITLE
Feat(web-twig): Introduce Modal Composed with subcomponents #DS-507

### PR DESCRIPTION
**Pull request by crishpeen**

### CODE REVIEW

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{# Attributes #}
+{%- set _idAttr = _id ? 'id="' ~ _id ~ '"' : null -%}
+{%- set _ariaLabelledbyAttr = _id ? 'aria-labelledby="' ~ _titleId ~ '"' : null -%}
+
+{# Miscellaneous #}
+{%- set _classNames = [ _rootClassName, rootComposedClassName, _rootExpandOnMobileClassName, _class ] -%}
+{%- set _mainPropsWithoutId = props | filter((value, prop) => prop != 'id') -%}
+
+{# Deprecations #}
+{% deprecated 'ModalComposed: The "ModalComposed" component is deprecated, it will replace "Modal".' %}
```
Comment (literat) [inline packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:22]:
```suggestion
{% deprecated 'ModalComposed: The "ModalComposed" component is deprecated, it will renamed to "Modal".' %}
```

File: packages/web-twig/src/Resources/components/Modal/Modal.twig [hunk 2]
```diff
@@ -18,6 +18,9 @@
 {%- set _classNames = [ _rootClassName, _class ] -%}
 {%- set _mainPropsWithoutId = props | filter((value, prop) => prop != 'id') -%}
 
+{# Deprecations #}
+{% deprecated 'Modal: The "Modal" component is deprecated, it will be replaced by "ModalComposed".' %}
```
Comment (literat) [inline packages/web-twig/src/Resources/components/Modal/Modal.twig:22]:
```suggestion
{% deprecated 'Modal: The "Modal" component is deprecated, it will be replaced by implementation of "ModalComposed".' %}
```

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{# API #}
+{%- set props = props | default([]) -%}
+{%- set _class = props.class | default(null) -%}
+{%- set _id = props.id -%}
+{%- set _isExpandedOnMobile = props.isExpandedOnMobile | default(false) | boolprop -%}
```
[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:5]: 🇬🇧

File: packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set dropdownDefault.twig__1.html [hunk 4]
```diff
@@ -1,8 +1,13 @@
 <html><body>
```
Comment (literat) [inline packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set dropdownDefault.twig__1.html:1]: Updated snapshot of the Dropdown should not be part of this PR.

Comment (crishpeen) [reply packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set dropdownDefault.twig__1.html:1]: my bad, sorry :)

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+{# Deprecations #}
+{% deprecated 'ModalComposed: The "ModalComposed" component is deprecated, it will be renamed to "Modal".' %}
+
+<dialog
+    {{ mainProps(_mainPropsWithoutId) }}
+    {{ classProp(_classNames) }}
+    {{ _idAttr | raw }}
+    {{ _ariaLabelledbyAttr | raw }}
+>
+    <div class="{{ _dialogClassName }}">
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:30]: As per the latest updates, this becomes an `<article>`.

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+{# Attributes #}
+{%- set _idAttr = _id ? 'id="' ~ _id ~ '"' : null -%}
+{%- set _ariaLabelledbyAttr = _id ? 'aria-labelledby="' ~ _titleId ~ '"' : null -%}
+
+{# Miscellaneous #}
+{%- set _classNames = [ _rootClassName, rootComposedClassName, _rootExpandOnMobileClassName, _class ] -%}
+{%- set _mainPropsWithoutId = props | filter((value, prop) => prop != 'id') -%}
+
+{# Deprecations #}
+{% deprecated 'ModalComposed: The "ModalComposed" component is deprecated, it will be renamed to "Modal".' %}
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:22]: Deprecation loop 🙂

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig [hunk 7]
```diff
@@ -0,0 +1,33 @@
+{# API #}
+{%- set props = props | default([]) -%}
+{%- set _class = props.class | default(null) -%}
+{%- set _id = props.id -%}
+{%- set _isExpandedOnMobile = props.isExpandedOnMobile | default(false) -%}
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:5]: Oh, the opposite is true:

```suggestion
{%- set _isExpandedOnMobile = props.isExpandedOnMobile | default(true) -%}
```

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+{%- set _rootClassName = _spiritClassPrefix ~ 'Modal' -%}
+{%- set _rootComposedClassName = _spiritClassPrefix ~ 'Modal--composed' -%}
+{%- set _rootExpandOnMobileClassName = _isBlock ? _isExpandedOnMobile ~ 'Modal--expandOnMobile' : null -%}
+
+{# Attributes #}
+{%- set _idAttr = _id ? 'id="' ~ _id ~ '"' : null -%}
+{%- set _ariaLabelledbyAttr = _id ? 'aria-labelledby="' ~ _titleId ~ '"' : null -%}
+
+{# Miscellaneous #}
+{%- set _classNames = [ _rootClassName, rootComposedClassName, _rootExpandOnMobileClassName, _class ] -%}
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:18]:
```suggestion
{%- set _classNames = [ _rootClassName, _rootComposedClassName, _rootExpandOnMobileClassName, _class ] -%}
```

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+{%- set props = props | default([]) -%}
+{%- set _class = props.class | default(null) -%}
+{%- set _id = props.id -%}
+{%- set _isExpandedOnMobile = props.isExpandedOnMobile | default(false) -%}
+{%- set _titleId = props.titleId | default(null) -%}
+
+{# Class names #}
+{%- set _rootClassName = _spiritClassPrefix ~ 'Modal' -%}
+{%- set _rootComposedClassName = _spiritClassPrefix ~ 'Modal--composed' -%}
+{%- set _rootExpandOnMobileClassName = _isBlock ? _isExpandedOnMobile ~ 'Modal--expandOnMobile' : null -%}
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:11]: 🤔 

```suggestion
{%- set _rootExpandOnMobileClassName = _isExpandedOnMobile ? 'Modal--expandOnMobile' : null -%}
```

File: packages/web-twig/src/Resources/components/Modal/ModalFooter.twig [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+{# Class names #}
+{%- set _actionsClassName = _spiritClassPrefix ~ 'ModalFooter__actions' -%}
+{%- set _descriptionClassName = _spiritClassPrefix ~ 'ModalFooter__description' -%}
+{%- set _rootAlignClassName = _align ? _spiritClassPrefix ~ 'ModalFooter--' ~ _align : null -%}
+{%- set _rootClassName = _spiritClassPrefix ~ 'ModalFooter' -%}
+
+{# Miscellaneous #}
+{%- set _classNames = [ _rootClassName, _rootAlignClassName, _class ] -%}
+
+<div
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalFooter.twig:16]: As per the latest updates, this becomes a `<footer>`.

File: packages/web-twig/src/Resources/components/Modal/ModalFooter.twig [hunk 11]
```diff
@@ -0,0 +1,28 @@
+{# API #}
+{%- set props = props | default([]) -%}
+{%- set _align = props.align | default(null) -%}
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalFooter.twig:3]:
```suggestion
{%- set _align = props.align | default('right') -%}
```

File: packages/web-twig/src/Resources/components/Modal/ModalHeader.twig [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        aria-controls="{{ _modalId }}"
+        aria-expanded="false"
+        color="tertiary"
+        data-dismiss="modal"
+        data-target="{{ '#' ~ _modalId }}"
+        isSquare
+    >
+        <Icon name="close" />
+        <span class="accessibility-hidden">{{ _closeLabel }}</span>
+    </Button>
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/ModalHeader.twig:41]: That's a question. I'm afraid we will be forced to turn this into a composition so users can reach the `Button` itself.

Since we are in Twig, it could be a block with a default content:

```twig
{% block content %}
  <Button …></Button>
{% endblock %}
```

But that might imply we need also `ModalTitle`…

@literat?

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 13]
```diff
@@ -1,7 +1,187 @@
...
+```
+
+👉 Please note the `titleId` attribute is linked to the title inside
+[Modal Header](#modalheader) and provides an accessible name for the dialog.
+
+#### Expand on Mobile Screens
+
+We recommend to expand the Modal on mobile screens using the
+`isExpandedOnMobile` prop. If you omit the prop, the Modal shrinks
+to fit the height of its content (if smaller than viewport).
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/README.md:39]: The recommendation is meant for the `web-twig` package. We can make it default in Twig and React.

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 14]
```diff
@@ -1,7 +1,187 @@
...
+
+### ModalFooter
+
+#### Examples:
+
+```twig
+<ModalFooter>
+  …
+</ModalFooter>
+<ModalFooter align="right" description="Modal Footer description">
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/README.md:121]: `right` is the default:

```suggestion
<ModalFooter description="Modal Footer description">
```

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 15]
```diff
@@ -1,7 +1,187 @@
...
+  …
+</ModalFooter>
+```
+
+#### API
+
+| Prop name     | Type     | Default | Required | Description                 |
+| ------------- | -------- | ------- | -------- | --------------------------- |
+| `align`       | `string` | `null`  | no       | Alignment of Footer Actions |
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/README.md:130]:
```suggestion
| `align`       | `string` | `right`  | no       | Alignment of Footer Actions |
```

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 16]
```diff
@@ -1,7 +1,187 @@
...
+```twig
+<button
+  type="button"
+  data-toggle="modal"
+  data-target="#modal-composed"
+  aria-controls="modal-composed"
+  aria-expanded="false"
+>
+  Open Modal
+</button>
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Modal/README.md:151]: We have `Button` in Twig 🎉

File: packages/web-twig/src/Resources/twig-components/modalFooterActions.twig [hunk 17]
```diff
@@ -0,0 +1 @@
+{% extends '@spirit/Modal/ModalFooterActions.twig' %}
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/twig-components/modalFooterActions.twig:1]: This file is redundant, is that right?

File: packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set modalComposedDefault.twig__1.html [hunk 18]
```diff
@@ -0,0 +1,52 @@
+<html><body>
```
Comment (adamkudrna) [inline packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set modalComposedDefault.twig__1.html:1]: Oh, this snapshot is a little bit messy… 🙈 But the other one looks correct to me 🙂.

**Review by adamkudrna — CHANGES_REQUESTED** [review]
Nice, however not there yet 🙂

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig (hunk 6 above)
Comment (crishpeen) [reply packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:22]: 🤯

File: packages/web-twig/src/Resources/components/Modal/ModalComposed.twig (hunk 9 above)
Comment (crishpeen) [reply packages/web-twig/src/Resources/components/Modal/ModalComposed.twig:11]:
```suggestion
{%- set _rootExpandOnMobileClassName = _isExpandedOnMobile ? _spiritClassPrefix ~ 'Modal--expandOnMobile' : null -%}
```
🤦

File: packages/web-twig/src/Resources/components/Modal/ModalHeader.twig (hunk 12 above)
Comment (literat) [reply packages/web-twig/src/Resources/components/Modal/ModalHeader.twig:41]: Is the close button a mandatory part of the design? E.g. if yes we should have it here and maybe in the future it will be the `CloseButton` component. If not, so we should apply default content.

But what is the case of reaching a button by developer? @adamkudrna

File: packages/web-twig/tests/components-fixtures/modalComposedDefault.twig [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        </ModalFooter>
+    </ModalDialog>
+</ModalComposed>
+
+<ModalComposed
+    id="modal-composed-2"
+    isExpandedOnMobile=false
+    titleId="modal-composed-2-title"
+>
+    <ModalDialog elementType="form" autocomplete="on" method="GET">
```
Comment (adamkudrna) [inline packages/web-twig/tests/components-fixtures/modalComposedDefault.twig:46]:
```suggestion
    <ModalDialog elementType="form" autocomplete="on" method="dialog">
```

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 20]
```diff
@@ -1,12 +1,241 @@
...
 
+## Composed Modal (DEPRECATED)
+
 This is Twig implementation of the [Modal] component.
 
+Modal is a composition of several subcomponents:
+
+- [ModalComposed](#modalcomposed)
+  - [ModalDialog](#modaldialog)
+  - [ModalHeader](#modalheader)
```
Comment (literat) [inline packages/web-twig/src/Resources/components/Modal/README.md:11]: The Header, Body, and Footer should not be on the same level as Dialog, should they?

File: packages/web-twig/src/Resources/components/Modal/README.md [hunk 21]
```diff
@@ -1,12 +1,241 @@
...
+| `id`      | `string` | —       | yes      | Modal ID                           |
+| `titleId` | `string` | `null`  | no       | ID of the title inside ModalHeader |
+
+On top of the API options, you can add `data-*` or `aria-*` attributes to
+further extend component's descriptiveness and accessibility. These attributes
+will be passed to the topmost HTML element of the component.
+
+### ModalDialog
+
+ModalDialog is the actual dialog window, a place for the header, body, and
```
Comment (literat) [inline packages/web-twig/src/Resources/components/Modal/README.md:50]: What about adding a link to the example with the form method dialog in the web package?

**Review by literat — APPROVED** [review]

File: packages/web-twig/src/Resources/components/Modal/README.md (hunk 21 above)
Comment (crishpeen) [reply packages/web-twig/src/Resources/components/Modal/README.md:50]: It is slightly different in twig, so I have the implementation explained here too.
